nifi-nar-bundles/nifi-standard-services/nifi-dbcp-service-bundle/nifi-dbcp-service/src/main/java/org/apache/nifi/dbcp/DBCPConnectionPool.java

```diff
@@ -460,45 +455,21 @@ public void onConfigured(final ConfigurationContext context) throws Initializati
         });
     }
 
-    private Long extractMillisWithInfinite(PropertyValue prop) {
-        return "-1".equals(prop.getValue()) ? -1 : prop.asTimePeriod(TimeUnit.MILLISECONDS);
+    private Driver getDriver(final String driverName, final String url) {
+        try {
+            final Class<?> clazz = Class.forName(driverName);
+            final Driver driver = DriverManager.getDriver(url);
+
+            return (driver == null)
```

Comment:
`DriverManager` never returns `null` but throws `SQLException` if no suitable driver found.
To handle that case, the exception handling needs to be improved, like this:
- try-catch for `Class.forName()` and handle `ClassNotFoundException` as failure
- try-catch for `DriverManager.getDriver()` and handle `SQLException` as falling back to direct instantiation
- try-catch for `clazz.newInstance()` and handle `IllegalAccessException` and `InstantiationException` as failure

However, this complicated error handling is not needed in my opinion. JDBC compliant drivers must register themselves and if the provided jdbc url is correct, then `DriverManager.getDriver(url)` should find the driver.
So the following would simply be enough:
```
Class.forName(driverClass);
return DriverManager.getDriver(jdbcUrl);
```
Though I do not insist on it. The fallback mechanism can be implemented just for sure.

Additional info from JDBC spec (http://download.oracle.com/otn-pub/jcp/jdbc-4_1-mrel-spec/jdbc4.1-fr-spec.pdf):

> JDBC drivers must implement the Driver interface, and the implementation must contain a static initializer that will be called when the driver is loaded. This initializer registers a new instance of itself with the DriverManager.

Reply:
Your concerns are valid, however it is hard to determine if the SQLException comes due to the driver implementation is not registering the driver instance as it is expected or because of the URL is incorrect. In both cases, the manager will not be able to return with a driver. But, as I said your concerns are valid, so I will go with the following implementation:

1. Checking if DriverManager.getDriver() returns with a driver. If yes, we will use that.
2. In case of SQLException, we try to remediate the situation with explicitly creating AND registering a driver instance
3. We try with getDriver() again. If it returns with a driver we can use that. Otherwise, if even after explicitly registering a driver, there is no sufficient one to return, then it is safe to assume that the URL is not supported by the driver.

I go with this, I hope this also covers your expectations